cmd/controller/apps_reconciler.go

```diff
@@ -35,5 +39,41 @@ func (r *AppsReconciler) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, err
 	}
 
-	return r.appFactory.NewCRDApp(existingApp, log).Reconcile()
+	crdApp := r.appFactory.NewCRDApp(existingApp, log)
+	r.UpdateAppRefs(crdApp.ResourceRefs(), existingApp)
+
+	force := false
+	if r.appUpdateStatus.IsUpdateNeeded(existingApp.Name, existingApp.Namespace) {
+		r.appUpdateStatus.MarkUpdated(existingApp.Name, existingApp.Namespace)
+		force = true
+	}
+
+	return crdApp.Reconcile(force)
+}
+
+func (r *AppsReconciler) UpdateAppRefs(refKeys map[reftracker.RefKey]struct{}, app *v1alpha1.App) {
+	// If App is being deleted, remove the App
+	// from all its associated references.
+	if app.DeletionTimestamp != nil {
+		r.appRefTracker.RemoveAppFromAllRefs(app.Name, app.Namespace)
+		return
+	}
+
+	// Make sure refs for App are always up to date
+	// in appRefTracker.
+	for refKey := range refKeys {
+		r.appRefTracker.AddAppForRef(refKey, app.Name)
+	}
+
+	// Remove any reference associations for App
+	// if the App no longer uses refs it once did.
+	r.appRefTracker.PruneAppFromRefs(refKeys, app.Name, app.Namespace)
```

Comment:
seems like it would be good to merge AddAppForRef and PruneAppFromRefs together since they are trying to "reconcile" refs (both are working on refKeys)

Reply:
Combined the two funcs as ReconcileRefs()